unixfs/hamt/fetcher.go

```diff
@@ -0,0 +1,297 @@
+	res  result
+}
+
+type jobStack struct {
+	c []*job
+}
+
+func (f *fetcher) mainLoop() {
+	var want *Shard
+	f.start = time.Now()
+	for {
+		select {
+		case id := <-f.reqRes:
+			if want != nil {
+				// programmer error
+				panic("fetcher: can not request more than one result at a time")
+			}
+			j, ok := f.jobs[id]
+			if !ok {
+				j = f.mainLoopAddJob(id)
+				if j == nil {
+					// no children that need to be retrieved
+					f.result <- result{vals: make(map[string]*Shard)}
+				}
+				if f.idle {
+					f.launch()
+				}
+			}
+			if j.res.vals != nil {
+				f.hits++
+				f.mainLoopSendResult(j)
+			} else {
+				if j.idx != -1 {
+					f.misses++
+					// move job to todoFirst so that it will be done on the
+					// next batch job
+					f.todo.remove(j)
+					f.todoFirst = j
+				} else {
+					f.nearMisses++
+				}
+				want = id
+			}
+		case bj := <-f.done:
+			f.doneCnt += len(bj.jobs)
+			f.cidCnt += len(bj.cids)
+			f.launch()
+			if want != nil {
+				j := f.jobs[want]
+				if j.res.vals != nil {
+					f.mainLoopSendResult(j)
+					want = nil
+				}
+			}
+			log.Infof("fetcher: batch job done")
+			f.mainLoopLogStats()
+		case <-f.ctx.Done():
+			log.Infof("fetcher: exiting")
+			f.mainLoopLogStats()
+			return
+		}
+	}
+}
+
+func (f *fetcher) mainLoopAddJob(hamt *Shard) *job {
+	children := hamt.missingChildShards()
+	if len(children) == 0 {
+		return nil
+	}
+	j := &job{id: hamt, cids: children}
+	if len(j.cids) > batchSize {
+		// programmer error
+		panic("job size larger than batchSize")
+	}
+	f.todo.push(j)
+	f.jobs[j.id] = j
+	return j
+}
+
+func (f *fetcher) mainLoopSendResult(j *job) {
+	f.result <- j.res
+	delete(f.jobs, j.id)
+	f.doneCnt--
+	if len(j.res.errs) != 0 {
+		return
+	}
+	// we want the first child to be the first to run not the last so
+	// add the jobs in the reverse order
+	for i := len(j.cids) - 1; i >= 0; i-- {
+		hamt := j.res.vals[string(j.cids[i].Bytes())]
+		f.mainLoopAddJob(hamt)
+	}
+	if f.idle {
+		f.launch()
+	}
+}
+
+func (f *fetcher) mainLoopLogStats() {
+	log.Infof("fetcher stats (cids, done, hits, nearMisses, misses): %d %d %d %d %d", f.cidCnt, f.doneCnt, f.hits, f.nearMisses, f.misses)
+	elapsed := time.Now().Sub(f.start).Seconds()
+	log.Infof("fetcher perf (cids/sec, jobs/sec) %f %f", float64(f.cidCnt)/elapsed, float64(f.doneCnt+f.hits+f.nearMisses+f.misses)/elapsed)
+}
+
+type batchJob struct {
+	cids []*cid.Cid
+	jobs []*job
+}
+
+func (b *batchJob) add(j *job) {
+	b.cids = append(b.cids, j.cids...)
+	b.jobs = append(b.jobs, j)
+	j.idx = -1
+}
+
+func (f *fetcher) launch() {
+	bj := batchJob{}
+
+	// always do todoFirst
+	if f.todoFirst != nil {
+		bj.add(f.todoFirst)
+		f.todoFirst = nil
+	}
+
+	// pop requets from todo list until we hit the batchSize
+	for !f.todo.empty() && len(bj.cids)+len(f.todo.top().cids) <= batchSize {
+		j := f.todo.pop()
+		bj.add(j)
+	}
+
+	if len(bj.cids) == 0 {
+		if !f.idle {
+			log.Infof("fetcher: entering idle state: no more jobs")
+		}
+		f.idle = true
```

Comment:
Possibly.  Although the current code has the limitation that there is only one batch job running at a time.  This can cause stragglers in a batch job to reduce throughput.  One ideas I explored earlier was to launch a new batch job when an existing job was half done.  I rejected this for now because the logic become to complicated.  (For example if we unconditional launch a new job when a batch is half done we might end up with a a large number of outstating jobs, we could limit the jobs also, but that complicated the logic even more.).  I would still like to explore this option at a later date.  I think I will change `idle` to `numBatchJobs` and make it clear that in the current iteration it is either 0 or 1, but at some point this could change.

Reply:
On second thought I think for this p.r. I will leave things as they are. I have reviewed the `idle` logic and am fairly certain it is correct.